drivers/net/dsa/lantiq_gsw_core.c

```diff
@@ -1817,7 +1801,9 @@ int gsw_core_probe(struct platform_device *pdev)
 		}
 	}
 
-	/* bring up the mdio bus */
+	/* TODO WARP-5828:
+	 * determine if this is applicable, move to platform probe if not.
```

Comment:
Probably can move to the platform probe. It'll be looking for a child node in the device tree that specifies the information about where to find a MDIO controlled slave. With our external part we won't be defining any child MDIO nodes. If we ever have extra PHY chips on the motherboard we'd have them defined as nodes on the same level because our chip would control them through the same or different means. Then connect them through links in the device tree. I think this is a good example of what that would look like https://elixir.bootlin.com/linux/v4.5/source/Documentation/devicetree/bindings/net/dsa/dsa.txt and page 38 of this (2 versus 3 switches) https://bootlin.com/pub/conferences/2018/elc/belloni-switchdev/belloni-switchdev.pdf

Reply:
Referencing the lantiq example here https://elixir.bootlin.com/linux/v5.10.119/source/Documentation/devicetree/bindings/net/dsa/lantiq-gswip.txt#L40 we'll probably be defining the structure a bit differently?

Reply:
I actually looked at this a little more today while writing some test code and do think this remains applicable (and that it will work!). The GSW parts have a "2nd" MDIO bus that they use to communicate with the internal (& external) PHYs connected to them (i.e. they are the "master" on that bus, where as on the one connected to the processor they are a "slave").

In order to expose those internal PHYs to the kernel we need to create another MDIO bus and provide read/write accessor functions for it (which will take comms destined for those PHYs and perform the necessary indirect addressing reads & writes to get the part to proxy the requests out onto it's internal MDIO bus). I think we can use the same DT node structure as the Lantiq doc file you linked.

In the future if we ever add additional external PHYs, you're right that we could just attach them to the "main" MDIO bus and skip all of the translation necessary to talk out over the GSW's secondary bus.

Reply:
Hmm... It's actually a bit of an open question whether this is necessary.

The part exposes registers to configure the internal PHYs, but also states they can be configured via the internal MDIO bus in compliance with the normal 802.3 MDIO stuff.

Hopefully the internally available registers will be all we need? I don't think it hurts to plan to bring this feature along though, at this point I don't think it will take much (if any) extra effort to get working.

Reply:
If our part has the internal MDIO, it's probably good to leave it in. Doesn't mean we'll test it though, so maybe leave a comment inidicating so